brayns/network/entrypoints/InspectEntrypoint.cpp

```diff
@@ -21,10 +21,103 @@
 
 #include "InspectEntrypoint.h"
 
+namespace
+{
+class SceneInspector
+{
+public:
+    SceneInspector(brayns::Engine &engine, const brayns::Vector2f &inspectPosition)
+        : _engine(engine)
+        , _inspectPosition(inspectPosition)
+    {
+    }
+
+    brayns::InspectResult inspect()
+    {
+        brayns::InspectResult result;
+
+        auto pickResult = _pickOSPRayScene();
+        if (!pickResult.hasHit)
+        {
+            result.hit = false;
+            return result;
+        }
+
+        auto &instance = _findHittedModel(pickResult.instance);
+        auto &model = instance.getModel();
+
+        result.hit = true;
+        result.model_id = instance.getID();
+
+        auto &ospHitPosition = pickResult.worldPosition;
+        result.position = brayns::Vector3f(ospHitPosition[0], ospHitPosition[1], ospHitPosition[2]);
+
+        const auto &hitPosition = result.position;
+        auto geometricModel = pickResult.model;
+        auto primitiveID = pickResult.primID;
+
+        brayns::JsonObject data;
+        model.onInspect(hitPosition, geometricModel, primitiveID, data);
+
+        auto &metadata = result.metadata;
+        metadata = brayns::JsonValue(data);
+
+        return result;
+    }
+
+private:
+    OSPPickResult _pickOSPRayScene()
+    {
+        auto &frameBuffer = _engine.getFrameBuffer();
+        auto &renderer = _engine.getRenderer();
+        auto &camera = _engine.getCamera();
+        auto &scene = _engine.getScene();
+
+        auto frameBufferHandle = frameBuffer.handle();
+        auto rendererHandle = renderer.handle();
+        auto cameraHandle = camera.handle();
+        auto sceneHandle = scene.handle();
+
+        auto x = _inspectPosition.x;
+        auto y = _inspectPosition.y;
+
+        OSPPickResult result;
+        ospPick(&result, frameBufferHandle, rendererHandle, cameraHandle, sceneHandle, x, y);
+        return result;
+    }
+
+    brayns::ModelInstance &_findHittedModel(OSPInstance hittedInstance)
+    {
+        auto &scene = _engine.getScene();
+        auto &models = scene.getModels();
+        auto &instances = models.getAllModelInstances();
+
+        auto begin = instances.begin();
+        auto end = instances.end();
+        auto instanceIterator = std::find_if(
+            begin,
+            end,
+            [&](brayns::ModelInstance *instancePtr) { return instancePtr->handle() == hittedInstance; });
+
+        // Shouldn't happen, but..
+        if (instanceIterator == end)
+        {
+            throw brayns::JsonRpcException("Could not find the hitted model instance in the scene");
+        }
```

Comment:
I have defined a lot of standard exceptions in JsonRpcException.h, I think you can use InternalError or something like that.